src/main/java/com/commercetools/sync/taxcategories/TaxCategorySync.java

```diff
@@ -0,0 +1,307 @@
+     *         batch to the target project.
+     */
+    @Override
+    protected CompletionStage<TaxCategorySyncStatistics> processBatch(@Nonnull final List<TaxCategoryDraft> batch) {
+        final Set<TaxCategoryDraft> validTaxCategoryDrafts = batch.stream()
+            .filter(this::validateDraft)
+            .collect(toSet());
+
+        if (validTaxCategoryDrafts.isEmpty()) {
+            statistics.incrementProcessed(batch.size());
+            return completedFuture(statistics);
+        } else {
+            final Set<String> keys = validTaxCategoryDrafts.stream().map(TaxCategoryDraft::getKey).collect(toSet());
+            return taxCategoryService
+                .fetchMatchingTaxCategoriesByKeys(keys)
+                .handle(ImmutablePair::new)
+                .thenCompose(fetchResponse -> {
+                    Set<TaxCategory> fetchedTaxCategories = fetchResponse.getKey();
+                    final Throwable exception = fetchResponse.getValue();
+
+                    if (exception != null) {
+                        final String errorMessage = format(TAX_CATEGORY_FETCH_FAILED, keys);
+                        handleError(errorMessage, exception, keys.size());
+                        return completedFuture(null);
+                    } else {
+                        return syncBatch(fetchedTaxCategories, validTaxCategoryDrafts);
+                    }
+                })
+                .thenApply(ignored -> {
+                    statistics.incrementProcessed(batch.size());
+                    return statistics;
+                });
+        }
+    }
+
+    /**
+     * Checks if a draft is valid for further processing. If so, then returns {@code true}. Otherwise handles an error
+     * and returns {@code false}. A valid draft is a {@link TaxCategoryDraft} object that is not {@code null} and its
+     * key is not empty.
+     *
+     * @param draft nullable draft
+     * @return boolean that indicate if given {@code draft} is valid for sync
+     */
+    private boolean validateDraft(@Nullable final TaxCategoryDraft draft) {
+        if (draft == null) {
+            handleError(TAX_CATEGORY_DRAFT_IS_NULL, null, 1);
+        } else if (isBlank(draft.getKey())) {
+            handleError(TAX_CATEGORY_DRAFT_HAS_NO_KEY, null, 1);
+        } else {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Given a {@link String} {@code errorMessage} and a {@link Throwable} {@code exception}, this method calls the
+     * optional error callback specified in the {@code syncOptions} and updates the {@code statistics} instance by
+     * incrementing the total number of failed tax categories to sync.
+     *
+     * @param errorMessage The error message describing the reason(s) of failure.
+     * @param exception    The exception that called caused the failure, if any.
+     * @param failedTimes  The number of times that the failed tax categories counter is incremented.
```

Comment:
handleError is called per draft/resource so I wonder why failedTimes might be > 1?

Reply:
Change to failedTimes increment by 1 every time invoking handleError
https://github.com/commercetools/commercetools-sync-java/pull/524/commits/9f2ff17c2139b63d53d8c305bfa53865e7b6edb6

Reply:
@leungkinghin did you address this comment ?

Reply:
Here we still need failedTime > 1 when several keys involves fetching service and fails
https://github.com/commercetools/commercetools-sync-java/blob/f4c5e6649daf0b257c48c5fd0ea3b840f64d3e45/src/main/java/com/commercetools/sync/taxcategories/TaxCategorySync.java#L110